examples/bridge-app/linux/include/Clusters.h

```diff
@@ -0,0 +1,371 @@
+
+// This is the interface to cluster implementations, providing access to manipulate attributes.
+class ClusterImpl
+{
+public:
+    virtual ~ClusterImpl() = default;
+
+    virtual void SetEndpointId(chip::EndpointId id)                                                                          = 0;
+    virtual chip::ClusterId GetClusterId()                                                                                   = 0;
+    virtual EmberAfStatus Read(const EmberAfAttributeMetadata * attributeMetadata, uint8_t * buffer, uint16_t maxReadLength) = 0;
+    virtual EmberAfStatus Write(const EmberAfAttributeMetadata * attributeMetadata, uint8_t * buffer)                        = 0;
+    virtual void WriteFromBridge(chip::AttributeId attributeId, const uint8_t * buffer)                                      = 0;
+    virtual chip::Span<const EmberAfAttributeMetadata> GetAllAttributes()                                                    = 0;
+};
+
+// *Type classes are used by code generation to specialize storage for a particular underlying type.
+
+// This provides storage for a primitive binary type, eg uintN_t, float, double
+// Values in ValueType are native, values in byte buffers are LE.
+template <typename T, size_t Bytes, EmberAfAttributeType AfType>
+struct PrimitiveType
+{
+    static constexpr EmberAfAttributeType kMatterType = AfType;
+    // Note that Bytes can be < sizeof(T) due to non-power-of-two byte lengths.
+    static constexpr uint16_t kMaxSize = Bytes;
+    static_assert(Bytes <= sizeof(T), "Incorrect type / byte specification");
+    typedef T ValueType;
+
+    PrimitiveType() = default;
+    PrimitiveType(ValueType initial) : mValue(initial) {}
+
+    EmberAfStatus Read(uint8_t * buffer, uint16_t sz) const
+    {
+        if (sz != kMaxSize)
+            return EMBER_ZCL_STATUS_FAILURE;
+        memcpy(buffer, &mValue, kMaxSize);
+        return EMBER_ZCL_STATUS_SUCCESS;
+    }
+
+    bool Equal(const uint8_t * buffer) const { return !memcmp(buffer, &mValue, kMaxSize); }
+
+    void Write(const uint8_t * buffer) { memcpy(&mValue, buffer, kMaxSize); }
+
+    bool TryWrite(ValueType newV)
+    {
+        if (mValue == newV)
+            return false;
+        mValue = newV;
+        return true;
+    }
+    bool TryWrite(const uint8_t * buffer)
+    {
+        if (Equal(buffer))
+            return false;
+        Write(buffer);
+        return true;
+    }
+
+    ValueType Peek() const { return mValue; }
+
+    uint8_t * GetBytes() { return reinterpret_cast<uint8_t *>(&mValue); }
+
+private:
+    ValueType mValue;
+    static_assert(std::is_standard_layout<ValueType>::value, "PrimitiveType not standard layout!");
+};
+
+// This provides access to a struct. It is similar to PrimitiveType except it is assumed
+// that each member handes its own endianness, instead of the whole type.
+template <typename T>
+struct StructType
+{
+    static constexpr EmberAfAttributeType kMatterType = ZCL_STRUCT_ATTRIBUTE_TYPE;
```

Comment:
I don't think we should be coupling the generation of metadata with the a specific interface for reading and writing. Similar to the precedence we already have for generating cluster objects (see `cluster-objects.h`) across all clusters and letting logic 'compose in' the ones they need to use for a given application setting, we should do something similar for attribute metadata.